### PR TITLE
[RFC] Linting: Linting: Recommend os_* instead of POSIX functions.

### PR DESCRIPTION
**Pull request by fwalch**
`rand_r` is deprecated in POSIX 2008.

### CODE REVIEW

**Comment by justinmk** [comment]
What about the other `_r` functions? Are they deprecated?

Does it make sense to have an os_rand() function?

**Comment by fwalch** [comment]
> What about the other _r functions? Are they deprecated?

Looks like some of them are. I'm not sure how to proceed here: 

 * remove the recommendations completely (e.g. remove `'rand(', 'rand_r('`)
    * maybe even don't check any of these thread-safety functions (https://github.com/neovim/neovim/pull/1758#issuecomment-69212292, https://github.com/neovim/neovim/pull/1493#discussion_r22246190)
 * recommend non-deprecated functions instead (e.g. say `'rand(', 'drand48_r('`)
 * add checks for deprecated functions (e.g. `'rand_r(', 'drand48_r('`)

> Does it make sense to have an os_rand() function?

I found no uses of `rand` and only one use of `rand_r` (`sha256.c`) in the codebase, so maybe not.

**Comment by justinmk** [comment]
[msdn thread](https://social.msdn.microsoft.com/Forums/en-US/3263ab0b-da3f-4737-9d80-0788a5b426db/randr-identifier-not-found?forum=vcgeneral) suggests `rand()`. And `drand48_r()` is not available in MSVC. So `rand()` is fine on Windows.

I think our general approach should be to abstract these cases in an `os_` function. [os_localtime_r]( https://github.com/neovim/neovim/blob/cb86eca91f9bdffe8b0214664169093d41902415/src/nvim/os/time.c#L78-L94) demonstrates this nicely. The clint warning should be changed to say:

> use os_foo instead (or implement it; see os_localtime_r for an example)

File: clint.py [hunk 1]
```diff
@@ -1207,7 +1207,6 @@ def CheckForMultilineCommentsAndStrings(filename, clean_lines, linenum, error):
     ('getpwuid(', 'getpwuid_r('),
     ('gmtime(', 'gmtime_r('),
     ('localtime(', 'localtime_r('),
-    ('rand(', 'rand_r('),
```
Comment (justinmk) [inline clint.py]: We can remove this one case, since os_rand is probably overkill.

File: clint.py [hunk 2]
```diff
@@ -1207,7 +1207,6 @@ def CheckForMultilineCommentsAndStrings(filename, clean_lines, linenum, error):
     ('getpwuid(', 'getpwuid_r('),
     ('gmtime(', 'gmtime_r('),
     ('localtime(', 'localtime_r('),
```
Comment (justinmk) [inline clint.py]: `localtime_r` should be changed to `os_localtime_r`

**Comment by justinmk** [comment]
On line 1239 where it says "for improved thread safety." can you change that to "(if it is missing, consider implementing it)"

File: clint.py (hunk 1 above)
Comment (fwalch) [inline clint.py]: OTOH, we need something to replace that one `rand_r` call in `sha256.c` at some point. But should it be `os_rand` or `os_rand_r`?

Comment (fwalch) [inline clint.py]: Then again, the function where `rand_r` is used, `sha2_seed`, is not used by Neovim.

**Comment by fwalch** [comment]
Updated.

**Comment by justinmk** [comment]
I think we should chop the _r suffix too. What do you think @philix?

**Comment by philix** [comment]
> I think we should chop the _r suffix too. What do you think @philix?

Better leave it. The `-_r` suffix communicates how the function is used and people may expect a function like `os_localtime` to work as the the `localtime` they already know. Since this is not true people may get confused. Besides that, removing them would be more work :)

**Pull request merged**